master/addexptrigger.sql

```diff
@@ -0,0 +1,150 @@
+ALTER TABLE experiments ADD COLUMN IF NOT EXISTS seq bigint;
+CREATE SEQUENCE IF NOT EXISTS stream_experiment_seq START 1;
+
+-- trigger function to update sequence number on row modification
+-- this should fire BEFORE so that it can modify NEW directly.
+CREATE OR REPLACE FUNCTION stream_experiment_seq_modify() RETURNS TRIGGER AS $$
+BEGIN
+    NEW.seq = nextval('stream_experiment_seq');
+RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+--
+DROP TRIGGER IF EXISTS stream_experiment_trigger_seq ON experiments;
+CREATE TRIGGER stream_experiment_trigger_seq
+    BEFORE INSERT OR UPDATE OF
+    state, archived, progress, start_time, end_time, notes
+                     ON experiments
+                         FOR EACH ROW EXECUTE PROCEDURE stream_experiment_seq_modify();
+
+-- helper function to create exp jsonb object for streaming
+-- halp!
+CREATE OR REPLACE FUNCTION stream_experiment_notify(
+    before jsonb, beforework integer, after jsonb, afterwork integer
+) RETURNS integer AS $$
+DECLARE
+output jsonb = NULL;
+    temp jsonb = NULL;
+BEGIN
+    IF before IS NOT NULL THEN
+        temp = before || jsonb_object_agg('workspace_id', beforework);
+output = jsonb_object_agg('before', temp);
+END IF;
+    IF after IS NOT NULL THEN
+        temp = after || jsonb_object_agg('workspace_id', afterwork);
+        IF output IS NULL THEN
+            output = jsonb_object_agg('after', temp);
+ELSE
+            output = output || jsonb_object_agg('after', temp);
+END IF;
+END IF;
+    PERFORM pg_notify('stream_experiment_chan', output::text);
+    -- seems necessary I guess
+return 0;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Trigger function to NOTIFY the master of changes, for INSERT/UPDATE/DELETE.
+CREATE OR REPLACE FUNCTION stream_experiment_change() RETURNS TRIGGER AS $$
+DECLARE
+proj integer;
+work integer;
+BEGIN
+    IF (TG_OP = 'INSERT') THEN
+        proj = project_id from experiments where experiments.id = NEW.experiment_id;
```

Comment:
You should be able to simplify this into one line
```suggestion
        work = workspace_id from projects where project.id = NEW.project_id;
```